src/server/pkg/storage/chunk/reader.go

```diff
@@ -4,109 +4,84 @@ import (
-	r := &Reader{
-		ctx:  ctx,
-		objC: objC,
-		buf:  &bytes.Buffer{},
-	}
-	if len(f) > 0 {
-		r.f = f[0]
+func newReader(ctx context.Context, objC obj.Client, dataRefs ...*DataRef) *Reader {
+	return &Reader{
+		ctx:      ctx,
+		objC:     objC,
+		dataRefs: dataRefs,
 	}
-	return r
 }
 
-// NextRange sets the next range for the reader.
-func (r *Reader) NextRange(dataRefs []*DataRef) {
+// NextDataRefs sets the next data references for the reader.
+func (r *Reader) NextDataRefs(dataRefs []*DataRef) {
 	r.dataRefs = dataRefs
-	r.r = bytes.NewReader([]byte{})
-	r.len = 0
-	for _, dataRef := range dataRefs {
-		r.len += dataRef.SizeBytes
+}
+
+func (r *Reader) Iterate(f func(*DataReader) error) error {
+	for {
+		if len(r.dataRefs) == 0 {
+			return nil
+		}
+		r.curr = newDataReader(r.ctx, r.objC, r.dataRefs[0], r.curr)
+		if err := f(r.curr); err != nil {
+			return err
+		}
+		r.dataRefs = r.dataRefs[1:]
 	}
+	return nil
 }
 
-// Len returns the number of bytes left.
-func (r *Reader) Len() int64 {
-	return r.len
+type DataReader struct {
+	ctx     context.Context
+	objC    obj.Client
+	dataRef *DataRef
+	chunk   []byte
+	seed    *DataReader
 }
 
-// Read reads from the byte stream produced by the set of DataRefs.
-func (r *Reader) Read(data []byte) (int, error) {
-	var totalRead int
-	defer func() {
-		r.len -= int64(totalRead)
-		r.bytesBeforeSplit += int64(totalRead)
-	}()
-	for len(data) > 0 {
-		n, err := r.r.Read(data)
-		data = data[n:]
-		totalRead += n
-		if err != nil {
-			if err := r.nextDataRef(); err != nil {
-				return totalRead, err
-			}
-		}
+func newDataReader(ctx context.Context, objC obj.Client, dataRef *DataRef, seed *DataReader) *DataReader {
+	return &DataReader{
+		ctx:     ctx,
+		objC:    objC,
+		dataRef: dataRef,
+		seed:    seed,
 	}
-	return totalRead, nil
 }
 
-func (r *Reader) nextDataRef() error {
-	// If all DataRefs have been read, then io.EOF.
-	if len(r.dataRefs) == 0 {
-		if r.f != nil {
-			dataRefs, err := r.f()
-			if err != nil {
-				return err
-			}
-			r.NextRange(dataRefs)
-		} else {
-			return io.EOF
-		}
-	}
-	// Get next chunk if necessary.
-	if r.curr == nil || r.curr.Chunk.Hash != r.dataRefs[0].Chunk.Hash {
-		r.executeOnSplitFunc()
-		if err := r.readChunk(r.dataRefs[0].Chunk); err != nil {
-			return err
-		}
+func (dr *DataReader) DataRef() *DataRef {
+	return dr.dataRef
+}
+
+func (dr *DataReader) Get(w io.Writer) error {
+	if err := dr.getChunk(); err != nil {
+		return err
 	}
-	r.curr = r.dataRefs[0]
-	r.dataRefs = r.dataRefs[1:]
-	r.r = bytes.NewReader(r.buf.Bytes()[r.curr.OffsetBytes : r.curr.OffsetBytes+r.curr.SizeBytes])
-	return nil
+	data := dr.chunk[dr.dataRef.OffsetBytes : dr.dataRef.OffsetBytes+dr.dataRef.SizeBytes]
+	_, err := w.Write(data)
+	return err
 }
 
-func (r *Reader) readChunk(chunk *Chunk) error {
-	objR, err := r.objC.Reader(r.ctx, path.Join(prefix, chunk.Hash), 0, 0)
+func (dr *DataReader) getChunk() error {
+	// Use seed chunk if possible.
```

Comment:
When is it not possible, presumably when we don't have a seed chunk, but when does that happen?